### ISSUE TITLE
WSL2: Help with Emacs

### ISSUE THREAD

**Issue by tajmone** [thoni56/c-xrefactory#67]
@thoni56, in these days I wanted to setup C-Xrefactory under Windows + WSL.

I'm trying to work my way through the various scattered instructions snippets found in the README, but haven't yet managed to work my head around them — hopefully, once I've managed to set it all up and running I'll write a step-by-step tutorial for Windows using WSL2, that includes instructions on how to setup a _specific_ version of Emacs and a specific Windows X11 server, so end users don't have to read up all the various links.

It's been quite a while since my last attempt to do, mostly because of the lack of clear instructions. So I'll take this chance to also share some thoughts on general project in this Issue.

## WSL and Emacs GUI?

From what I understood, the Windows X11 server is only required for GUI versions of Emacs (i.e. not terminal based, e.g. via curses). I was wondering, isn't it possible to just install a Windows version of Emacs (GUI) and bridge it over to WSL to interface with XRefactory, instead of running it in WSL and bridging it over to the Windows X11 server running under Win OS? IRC, WSL allows bridging process over to Win OS, and since the protocol uses STD streams this should be possible (maybe it requires tweaking the ELisp scripts a bit).

I would really love to be able to use C-Xrefactory under Win OS, even if it means running it via WSL and using it on Emacs. Once I've managed to set it up and start using it, I could contribute to project via documentation and scripts for Win OS usage.

## Easy Installation Assets

I think that anything that makes it easier for end users to setup C-Xref under any OS is going to contribute to the popularity of this repository, which hopefully could translate into contributions to its growth.

For example, it would be really great if you could also add a `Makefile` for cross-compiling C-XRefactory to Windows binaries from Linux, and publish the pre-built binaries somewhere.

## Cross-Compiling Win Binaries?

In the "Setup" section you mention that it's possible to compile it under Windows using Cygwin or Msys2-MSYS:

> `c-xrefactory` works with recent Emacsen on Linux, including WSL, Cygwin, Msys2-MSYS (but not Msys2-MingW*) and MacOS/Darwin.

What is exactly "Msys2-MSYS" (as opposed to Msys2-MingW)? As far as I'm aware, the MSYS2 project (www.msys2.org) only offers the MingW variant.

I would like to cross-compile C-XRefactory to Windows from WSL, and if I succeed write a guide with step-by-step instruction and maybe provide the binaries myself. What should I use under WSL, **Cygwin** or **Msys2-MSYS**? if the latter allows creating a standalone binary without the Cygwin DLL dependency I'd rather use that. Any help in this direction is much appreciated.

# Other Editors' Support

I've been reading your Wiki documentation on how C-Xref interfaces with Emacs, via its custom protocol over STD stream. It looks like it should be possible to integrate C-Xref into other editors too, by either emulating the protocol or via an Elisp interpreter — which would still be a great solution in the meantime, until LSP will be supported.

I gather that the C-Xref server protocol is still mostly undocumented and obscure regarding the details of what it does, although you managed to create a spy tool to log all communications. In terms of interfacing editors, we really only need to focus on the major commands and update operations from the editor side, and how these are notified to C-Xref; wouldn't inspection of the ELisp extension suffice in this respect?

Failing all else, it should be possible to add to the editor plug-in an ELisp interpreter in order to emulate Emacs behaviour in respect to all communications with C-Xref. There are many ELisp compilers and interpreters that could be used for this end, e.g. [El Compilador] is an [SSA-based] compiler for Emacs Lisp (a project with the goal of porting Emacs sources from C to ELisp, making it a self-hosting project).

In theory, nothing prevents an LSP Language Server to use ELisp to interface with the protocol, if that does the work. It might add some overhead, and not be a clean solution, but it's still better than nothing while waiting that the XRef protocol is fully "cracked" and replaceable with a pure LSP compliant alternative.

Whatever the solution, being able to sue C-Xrefactory from other editors too would increase its potential userbase, and bring in contributions.

# Various Considerations

A few lateral considerations that somehow intersect what written earlier.

## OS Specific Assets

If I manage to come up with a guide and/or scripts to simplify setting up CXrefactory under Windows (via WSL2 or cross-compiling), where should these OS specific assets go?

Would you like to host them directly in the repository, in a dedicate folder, or what you rather have me create a dedicated repository for each asset?

## Documentation from Source via Generator

While reading my way through the docs in the `docs/` folder, the instructions in the README and the Wiki documentation, I couldn't avoid noticing that notes on the project status and TODOs are sort of scattered all around the place, yet the repository is not making much use of the Dashboard (Projects, Milestones and the new Dashboard Beta features which are great).

Wouldn't it make more sense to start documenting the project directly in the source code, using a language agnostic documentation tool, so whenever you work on a module or function your annotations are immediately reflected in the public documentation? this would save you the double work of having to manually update the Wiki to ensure that it's up to date, and would also make it much easier for contributors to either help documenting the project or contribute to the code (i.e. without having to worry about the Wiki).

An added benefit of using a documentation from source generator would be that you could serve the documentation via GHPages from the repository, and the documentation would always contain direct links to the actual repository sources.

[El Compilador]: https://github.com/tromey/el-compilador
[SSA-based]: https://en.wikipedia.org/wiki/Static_single_assignment_form

[RDoc]: https://ruby.github.io/rdoc/

**Comment by thoni56** [comment]
Most of what you propose has gone through my mind too. But at this point nothing of that will happen. Remember that this is a long standing, slowly progressing, resurrection project of legacy code that I work on for my own pleasure and learning. Even after 7 years I hardly understand what everything does or how to make the bigger changes that I, at this point, think I want to try at some future point.

I have a roadmap of what I intend to do, in time, and inspired of what I find during the untangling of the legacy code. This might include radical restructuring, removing features etc. so I don't actully what more "order". At this point.

I don't actually even want the user base to grow at this time.

Not every project on GitHub is a finished, usable product. If someone finds `c-xrefactory` and think it helps them I'm glad, but "user support" will be extremely limited.

With that said, a couple of points, or helping comments:

1. SETUP: The only supported way to set up `c-xref` is the one clearly described in the README: find an Emacs, install `el-get` and install the `c-xrefactory` package from that. Run the tutorial from the menu to learn how to use it. If you are using Emacs daily those instructions are sufficient, if not, your Emacs knowledge is probably not on the level that you should atempt to use `c-xrefactory` but instead try another alternative like Visual Studio Code with `clangd` lsp support.
2. ENVIRONMENTS: In Windows this works using Cygwin and Msys "native". WSL is so much faster that I've almost completely left the Cygwin environment (and I will never go down the road of Microsoft developer toolchains...).
3. MSYS2: When you install Msys2 you get access to different environments, the 64-bit Mingw, the 32-bit Mingw and the "native" (x86_64-pc-msys), which is basically the same as Cygwin. [Here is a list](https://www.msys2.org/docs/environments/). If you press the Windows key and type "Msys" you will see some choices, I don't remember if you need to activate them separately during install or using a `pacman` install.
4. INSTALL: You can simply install Emacs using the relevant package manager and follow the instructions in the README to get going. As long as you are in one of the environments where `c-xref` compiles. No need to cross-compile, elisp interpreters or other complicated stuff.
5. EMACS: Both Cygwin and Msys2 has text-version Emacs available, Cygwin, at least, also has a windowed version.
6. EDITORS: Re-implementing the editor adaptor is in itself a multi-year effort if you are not fluent in e-lisp and will be vulnarable to any changes in the entangled and complex interaction between `c-xref` and the editor adaptor. Again, in order to be able to ruthlessly refactor this system into something nice, this interface is not considered a stable, supported "API". With Emacs as the only supported editor and `el-get` as the only supported install method the c-xref proper and the editor adapter are guaranteed to be in sync without extra effort or problems.
7. DOCUMENTATION: It sounds intriguing to document in source code, I'm all for it (my first real project was building such a tool!), but you are only mentioning GitHub planning tools, so I'm uncertain what you actually mean. I suggest you mail me privately with some more concrete ideas to try. I'm not going to use GitHub issues and boards for planning.

I hope that helps somewhat, if you *really* are sure that you are interested in using `c-xrefactory` (as Visual Studio Code, or even Emacs, with clang lsp is a much more mature, stable and supported alternative for basic C cross-referencing and navigation).

I also hope it clears up what this project is, what state it is in and what support you can (can't) expect.

TLDR: This project is in a constant flux, and will be so for a long time to come.

**Comment by thoni56** [comment]
@tajmone Thanks for all your suggestions! Sorry to not be more supportive, I hope you find some of my tips to be helpful at least.

Re: documentation, I have been thinking about what would be helpful and I think (and it might actually be what you hinted at) that converting the wiki pages into a "normal" asciidoc document to keep in the source tree, that would really be nice.

I've already toyed with PlantUML/C4 models, as you can see in the `/design` folder.

So if you want to help out, I think that is an area that I'd gladly take your help. I know you are extremely good at that too. This would enable me to continuously grow the documentation of this along with the discovery and refactoring of it. Although I have the wiki repo checked out so I *can* update it (and just push to publish), it is a bit fiddly to work with.

**Comment by tajmone** [comment]
> @tajmone Thanks for all your suggestions! Sorry to not be more supportive, I hope you find some of my tips to be helpful at least.

I haven't been able to try them out yet but I should manage to (time permitting).

> Re: documentation, I have been thinking about what would be helpful and I think (and it might actually be what you hinted at) that converting the wiki pages into a "normal" asciidoc document to keep in the source tree, that would really be nice.

If I may drop a suggestion here, I strongly advise to look into the Zig language:

https://ziglang.org/

It's a new (Beta but almost reaching 1.0) language designed to be fully compatible with C and C++, Zig modules can interact with C both ways, and does it so good that you don't even have to create headers and bindings (simply linking to a C module exposes all its variables, functions and macros).

> Zig is better at using C libraries than C is at using C libraries.

Zig is so good in handling C and C++ that many C/C++ developers are using it as a replacement for C compilers. It's a LLVM based tool, and it includes all the known C ABIs for all the supported platforms, which allows to resolve compiler incompatibility issues at compile time, as well as allowing to cross-compile for all supported platforms from every supported Tier-1 platform — i.e. no more fiddling with Cygwin, MinGW, etc, since Zig can quickly transform any project into a fully cross platform codebase.

> Zig can build for any of the targets from the Support Table with Tier 3 Support or better. No “cross toolchain” needs to be installed or anything like that. 

Since Zig comes with its own self-documenting system (works also for C) just by using Zig you'd be solving the documentation problem.

Furthermore, Zig can translate C modules to Zig sources without any manual fixes (and for the edge cases which it can't handle, it simply keeps the C block as-is, which is still usable from within the Zig source):

- [Export functions, variables, and types for C code to depend on](https://ziglang.org/learn/overview/#export-functions-variables-and-types-for-c-code-to-depend-on)

For example, I've read your Blog article [_Refactoring macros to functions_](https://www.responsive.se/thomas/2020/05/14/refactoring-macros-to-functions/). This is [something which Zig can handle natively in many cases](https://ziglang.org/documentation/0.9.1/#toc-C-Macros):

> C Translation makes a best-effort attempt to translate function-like macros into equivalent Zig functions. Since C macros operate at the level of lexical tokens, not all C macros can be translated to Zig. Macros that cannot be translated will be be demoted to @compileError. Note that C code which uses macros will be translated without any additional issues (since Zig operates on the pre-processed source with macros expanded). It is merely the macros themselves which may not be translatable to Zig.

Many C developers today are using Zig to revive old legacy code because it solves many C compatibility issues thanks to its rich ABI libraries, and also because it makes easier to refactor old code.

It also solves many of the known C problems relating to undefined behaviour, in a fashion inspired by Rust but not as complicated. Zig is especially safe when it comes to pointers — of which it provides multiple variants — and memory management.

I think that Zig could help in this project, especially since it provides many useful native features:

1. Documentation from source comments (Zig and C/C++).
2. A complete [build system](https://ziglang.org/learn/overview/#zig-build-system), much better than `make` or other tools.
3. [Full testing capabilities](https://ziglang.org/documentation/0.9.1/#Zig-Test), allowing tests to be included in the source code.

Since the Zig compiler is based in LLVM, chances are that it could also help understand the more obscure parts of the original code, thanks to LLVM rich toolset.

You might want to give it a try, even if only to make use of its documentation and test system. Zig also has an LSP language server, called ZLS:

- https://github.com/zigtools/zls

which has also a client plug-in for Emacs:

- https://github.com/emacs-lsp/lsp-mode
- https://github.com/ziglang/zig-mode
